Make key binding syntax more consistent

krader1961 opened the pull request:
Document the format of key names. Make function key name matching
case-insensitive for consistency with how key modifers (e.g., "Alt")
are handled.

Resolves #1163

krader1961 commented:
I'm not entirely happy with this change since it illustrates why the (sometimes) case-insensitivity greatly complicates the documentation. This also doesn't properly document the behavior of the `Shift` modifier but that's for another day. But it does at least document the current behavior and makes `Ctrl-?` consistent with `"\c?"`.

krader1961 commented:
Also, I'll be happy to add unit tests for the changed behaviors if you're otherwise happy with simply documenting the current behavior and making function key names case insensitive to match how key modifiers are matched.

xiaq commented:
I'd rather make key bindings consistently case-sensitive, and provide an easy way to insert the name of a key in CLI.

krader1961 commented:
> I'd rather make key bindings consistently case-sensitive, ...

That's my preference as well. I only implemented the case insensitive solution because you had earlier implied a preference for it, but perhaps I misinterpreted what you wrote.

> ... and provide an easy way to insert the name of a key in CLI.

Is there any reason that enhancement needs to be coupled with making the case sensitivity consistent, fixing the handling of `Ctrl-?`, and documenting these symbols?

krader1961 commented:
FWIW, I modified my change to make the key modifier case sensitive -- rather than make function names case insensitive. That results in several unit test failures since they expect key modifiers to be case insensitive.

krader1961 commented:
@xiaq, PTAL. I've modified the change to make key modifiers case sensitive. Obviously a backward incompatible change so I've noted that in the release notes and modified the unit tests accordingly. I also opened issue #1180 to track the desirability of adding a mechanism for interactively inserting the symbolic name for a key.

krader1961 commented:
@xiaq Ping? This should be backward compatible with the exception of key modifiers no longer being case-insensitive. Which I have documented and is unlikely to cause problems for current Elvish users.

xiaq commented on NEXT-RELEASE.md:
```diff
@@ -11,6 +11,10 @@ This is the draft release notes for 0.15.0, scheduled to be released on
     -   The `rest-arg` field now contains the index of the rest argument,
         instead of the name.
 
+-   Key modifiers are no longer case insensitive. Thus `Alt` is still recognized
```
> Thus `Alt` is still recognized

I'd stick to "for example" for examples

xiaq commented on pkg/ui/key.go:
```diff
@@ -122,57 +140,64 @@ func (k Key) String() string {
 
 	if len(s) == 1 {
 		k.Rune = rune(s[0])
+		if k.Rune < ' ' {
```
Using the numerical value of 0x20 is a bit clearer here.

xiaq commented on pkg/ui/key.go:
```diff
@@ -122,57 +140,64 @@ func (k Key) String() string {
+			// Convert literal control char to the equivalent canonical form;
+			// e.g., "\e" to Ctrl-'[' and "\t" to Ctrl-I.
+			k.Mod |= Ctrl
+			k.Rune += '@'
```
Using the numerical value of 0x40 is much clearer than `@` here.

xiaq commented on pkg/ui/key.go:
```diff
@@ -182,33 +207,29 @@ func ParseKey(s string) (Key, error) {
 			} else if k.Rune == 'J' {
 				k.Mod &= ^Ctrl
 				k.Rune = Enter
+			} else if k.Rune == '?' {
+				k.Mod &= ^Ctrl
+				k.Rune = Backspace
```
I don't think typing `Ctrl-?` actually produces backspace in any modern terminal, so I'd rather leave this out.

xiaq commented on pkg/ui/key_test.go:
```diff
@@ -51,31 +51,44 @@ var parseKeyTests = []struct {
 	{s: "C-x", wantKey: Key{'X', Ctrl}},
 	{s: "C-X", wantKey: Key{'X', Ctrl}},
 
+	// Literal control chars are equivalent to the preferred Ctrl-? formulation.
```
> Ctrl-?

I'd use "Ctrl-<key>" instead as this can be misunderstood as the Ctrl-? key combination.

xiaq commented on pkg/ui/key_test.go:
```diff
@@ -51,31 +51,44 @@ var parseKeyTests = []struct {
 
+	// Literal control chars are equivalent to the preferred Ctrl-? formulation.
+	{s: "\033", wantKey: Key{'[', Ctrl}},
+	{s: "\t", wantKey: K(Tab)},
```
This is duplicate with a test below. Also, parsing of `\t` doesn't go through the code path for parsing control characters.

xiaq commented on pkg/ui/key_test.go:
```diff
@@ -51,31 +51,44 @@ var parseKeyTests = []struct {
+	{s: "\033", wantKey: Key{'[', Ctrl}},
+	{s: "\t", wantKey: K(Tab)},
+	{s: "Alt-\t", wantKey: Key{Tab, Alt}},
+	{s: "\n", wantKey: K(Enter)},
```
Same

xiaq reviewed (approved):
Mostly LGTM. A few small issues that I'll address after merging this.
